modyn/selector/internal/selector_strategies/coreset_strategy.py

```diff
@@ -0,0 +1,102 @@
+import random
+from typing import Iterable
+
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+from modyn.metadata_database.models import SelectorStateMetadata
+from modyn.selector.internal.selector_strategies import AbstractSelectionStrategy
+from modyn.selector.internal.selector_strategies.downsampling_strategies import (
+    AbstractDownsamplingStrategy,
+    EmptyDownsamplingStrategy,
+    instantiate_downsampler,
+)
+from modyn.selector.internal.selector_strategies.presampling_strategies import (
+    AbstractPresamplingStrategy,
+    EmptyPresamplingStrategy,
+    instantiate_presampler,
+)
+
+
+class CoresetStrategy(AbstractSelectionStrategy):
+    def __init__(self, config: dict, modyn_config: dict, pipeline_id: int, maximum_keys_in_memory: int):
+        super().__init__(config, modyn_config, pipeline_id, maximum_keys_in_memory)
+
+        self.presampling_strategy: AbstractPresamplingStrategy = instantiate_presampler(
+            config, modyn_config, pipeline_id, maximum_keys_in_memory
+        )
+
+        self.downsampling_strategy: AbstractDownsamplingStrategy = instantiate_downsampler(
+            config, maximum_keys_in_memory
+        )
+
+        if isinstance(self.presampling_strategy, EmptyPresamplingStrategy) and isinstance(
+            self.downsampling_strategy, EmptyDownsamplingStrategy
+        ):
+            raise ValueError(
+                "Using a Coreset Method without presampling and downsampling is useless. "
+                "You can use NewDataStrategy instead. "
+                "To specify the presampling method add 'presampling_strategy to the pipeline. "
+                "You can use 'downsampling_strategies' to specify the downsampling method."
+            )
+
+    def inform_data(self, keys: list[int], timestamps: list[int], labels: list[int]) -> None:
+        assert len(keys) == len(timestamps)
+        assert len(timestamps) == len(labels)
+
+        self._persist_samples(keys, timestamps, labels)
+
+    def _on_trigger(self) -> Iterable[list[tuple[int, float]]]:
+        for samples in self._get_data():
+            random.shuffle(samples)
+            yield [(sample, 1.0) for sample in samples]
+
+    def _get_data(self) -> Iterable[list[int]]:
+        """Returns all sample
+
+        Returns:
+            list[str]: Keys of used samples
+        """
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            if self.presampling_strategy.requires_trigger_dataset_size():
```

Comment:
Many presampling strategies require to know the full dataset size. To avoid code repetition, this bool function is added to the abstract class.